enterprise/cmd/repo-updater/internal/authz/perms_syncer.go

```diff
@@ -230,6 +233,56 @@ func (s *PermsSyncer) listPrivateRepoNamesBySpecs(ctx context.Context, repoSpecs
 	return repoNames, nil
 }
 
+func (s *PermsSyncer) refreshGitLabOAuthToken(ctx context.Context, acct *extsvc.Account) error {
+	if acct.ServiceType != extsvc.TypeGitLab {
+		return nil
+	}
+
+	var oauthConfig *oauth2.Config
+	for _, authProvider := range conf.SiteConfig().AuthProviders {
+		if authProvider.Gitlab == nil ||
+			strings.TrimSuffix(acct.ServiceID, "/") != strings.TrimSuffix(authProvider.Gitlab.Url, "/") {
+			continue
+		}
+		url := strings.TrimSuffix(authProvider.Gitlab.Url, "/")
+
+		oauthConfig = &oauth2.Config{
+			ClientID:     authProvider.Gitlab.ClientID,
+			ClientSecret: authProvider.Gitlab.ClientSecret,
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  url + "/oauth/authorize",
+				TokenURL: url + "/oauth/token",
+			},
+			Scopes: []string{"read_user", "api"},
+		}
+		break
+	}
+	if oauthConfig == nil {
+		return nil
+	}
+
+	_, tok, err := gitlab.GetExternalAccountData(&acct.AccountData)
+	if err != nil {
+		return errors.Wrap(err, "get external account data")
+	} else if tok == nil {
+		return errors.New("no token found in the external account data")
+	}
+
+	refreshedToken, err := oauthConfig.TokenSource(ctx, tok).Token()
+	if err != nil {
+		return errors.Wrap(err, "refresh token")
+	}
+
+	if refreshedToken.AccessToken != "" && refreshedToken.AccessToken != tok.AccessToken {
```

Comment:
Based on docs for `TokenSource`:
> TokenSource returns a TokenSource that returns t until t expires, automatically refreshing it as necessary using the provided context.

It seems to me that we can compare just `refreshedToken != tok` here:
```suggestion
	if tok != refreshedToken && refreshedToken.AccessToken != "" {
```